File: packages/canvas/src-native/canvas-android/canvas/src/main/java/com/github/triniwiz/canvas/LinearGradient.java
package com.github.triniwiz.canvas;

/**
 * Created by triniwiz on 5/30/20
 */
public class LinearGradient extends Gradient {
    float x0;
    float y0;
    float x1;
    float y1;

    public LinearGradient(float x0, float y0, float x1, float y1) {
        this.x0 = x0;
        this.y0 = y0;
        this.x1 = x1;
        this.y1 = y1;
    }
}

File: packages/canvas/Dom/shaders/LinearGradient.ts
import { Property } from '@nativescript/core';
import { Gradients } from './Gradients';

const startProperty = new Property<LinearGradient, { x: number; y: number }>({
	name: 'start',
});

const endProperty = new Property<LinearGradient, { x: number; y: number }>({
	name: 'end',
});

const colorsProperty = new Property<LinearGradient, string[]>({
	name: 'colors',
});

export class LinearGradient extends Gradients {
	start: { x: number; y: number };
	end: { x: number; y: number };
	colors: string[];

	_getColor() {
		const ctx = this._canvas.getContext('2d');
		const gradient = ctx.createLinearGradient(this.start.x, this.start.y, this.end.x, this.end.y);

		for (let i = 0; i < this.colors.length; i++) {
			gradient.addColorStop(i, this.colors[i]);
		}

		return gradient;
	}

	draw(): void {
		const ctx = this._canvas.getContext('2d');
		const gradient = this._getColor();
		if (this._getPaintStyle() === 'stroke') {
			ctx.strokeStyle = gradient;
		} else {
			ctx.fillStyle = gradient;
		}
	}
}

startProperty.register(LinearGradient);
endProperty.register(LinearGradient);
colorsProperty.register(LinearGradient);
